### PR TITLE
Added runtime TypeError for non-function, non-null __extends

### LINKED ISSUE #5794: Add explicit error to __extends when base class is undefined
We occasionally get people in the IRC channel asking about their transpiled JS failing with `TypeError: undefined is not an object`. It happens when accessing `b.prototype` inside `__extends` because the child class is before the base class, but they often think it's a bug in `__extends`. Would it make sense to detect if `b` is undefined at the start of the function and throw an appropriately worded error? Eg:

``` typescript
var __extends = this.__extends || function (d, b) {
    if (b === undefined) { throw new Error("The base class is not defined. Did you put the child class before the base class?"); }
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
```

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
As per the linked issue, although TypeScript already has checking in place that tries to prevent users from extending classes before their declaration, it's still possible to accidentally work around the checker. This adds a block to `__extends` that throws a `TypeError` if the base class `b` isn't a function.

This'll be what the new error message is:

```
C:\Code\tstest\index.js:10
            throw new TypeError("Class extends value " + b + " is not a constructor or null");
            ^

TypeError: Class extends value undefined is not a constructor or null
    at __extends (C:\Code\tstest\index.js:10:19)
    at C:\Code\tstest\index.js:18:5
    at Object.<anonymous> (C:\Code\tstest\index.js:23:2)
```

My _hope_ is that this will not have a negative performance impact on community performance-critical applications, as they would likely already prefer newer browser/Node versions and output ES2015+ code. If there is something you can think of that I should do to verify that hope, I'd love to know!

<details>
<summary>These errors match the behavior in Chrome, Node, and Firefox</summary>
Runtime errors in Chrome 80 and Node 12.0.0:

```js
class X extends null { }
// undefined

class Y extends undefined { }
// TypeError: Class extends value undefined is not a constructor or null

class Z extends 0 { }
// TypeError: Class extends value 0 is not a constructor or null
```

`Class extends value {0} is not a constructor or null` matches the Node.js behavior:
* [Message template](https://github.com/nodejs/node/blob/2bdeb88c27b4d8de3a8f6b7a438cf0bcb88fa927/deps/v8/src/common/message-template.h) for `ExtendsValueNotConstructor`
* [Error thrown with that message](https://github.com/nodejs/node/blob/6ca81ad72a3c6fdf16c683335be748f22aaa9a0d/deps/v8/src/runtime/runtime-classes.cc#L617) when `!super_class->IsConstructor()`

Runtime errors in Firefox 72.0.1:

```js
class X extends null { }
// undefined

class Y extends undefined { }
// TypeError: class heritage undefined is not an object or null

class Z extends 0 { }
// TypeError: class heritage 0 is not an object or null
```
</details>

Fixes #5794.

### CODE REVIEW

File: src/compiler/transformers/es2015.ts [hunk 1]
```diff
@@ -62,9 +62,9 @@ namespace ts {
     }
 
     const enum Jump {
-        Break       = 1 << 1,
```
Comment (JoshuaKGoldberg) [inline src/compiler/transformers/es2015.ts]: Agh. The whitespace changes are from me forgetting to Save Without Formatting. Will revert...

**Comment by JoshuaKGoldberg** [comment]
@typescript-bot perf test
@typescript-bot pack this

**Comment by JoshuaKGoldberg** [comment]
@typescript-bot perf test
@typescript-bot pack this

**Comment by JoshuaKGoldberg** [comment]
Well, I guess @typescript-bot only listens to Microsoft employees 😄

**Comment by sandersn** [comment]
Yep, sorry! Typescript-bot is (1) powerful (2) ruthless, so we don't trust anybody outside the team to command it.

@typescript-bot perf test this
@typescript-bot pack this

**Comment by rbuckton** [comment]
In general this looks fine, though we'd have to make a similar change to https://github.com/microsoft/tslib. One thing to note is that this does not behave the same as Chrome/v8 for this case:
```ts
class C extends Symbol.iterator {}
// v8: TypeError: Class extends value Symbol(Symbol.iterator) is not a constructor or null
// this emit: TypeError: Cannot convert a Symbol value to a string.
```

This could be addressed by changing the exception to this instead:
```ts
throw new TypeError("Class extends value " + String(b) + " is not a constructor or null.");
```

As `String(Symbol.iterator)` coerces symbols to strings using `.toString()`.

**Comment by sandersn** [comment]
@rbuckton should we merge this now or wait for the tslib change?

**Comment by rbuckton** [comment]
I am reviewing now. While we should still have a PR for tslib, there are no changes here that would make them incompatible.

**Review by rbuckton — APPROVED** [review]

**Comment by JoshuaKGoldberg** [comment]
Super, thanks for the reviews! I can send that PR over to tslib.

**Pull request merged**